addons/account/models/account.py

```diff
@@ -1869,6 +1869,66 @@ def _fix_tax_included_price_company(self, price, prod_taxes, line_taxes, company
+        """ Helper to get the price unit, may be called from different models
+        """
+        company = company or self.env.company
+        company_currency = company.currency_id
+        currency = currency or company_currency
+
+        if product_price_unit == None or product_taxes == None:
+            if document_type == 'sale':
+                product_price_unit = product.lst_price
+                product_taxes = product.taxes_id
+            elif document_type == 'purchase':
+                product_price_unit = product.standard_price
+                product_taxes = product.supplier_taxes_id
+            else:
+                return 0.0
+        product_taxes = product_taxes.filtered(lambda tax: tax.company_id == company)
+
+        # Apply unit of measure.
+        if product_uom and product.uom_id != product_uom:
+            product_price_unit = product.uom_id._compute_price(product_price_unit, product_uom)
+
+        # Apply fiscal position.
+        if product_taxes and fiscal_position:
+            product_taxes_after_fp = fiscal_position.map_tax(product_taxes, partner=partner)
+
+            if set(product_taxes.ids) != set(product_taxes_after_fp.ids):
+                flattened_taxes_before_fp = product_taxes._origin.flatten_taxes_hierarchy()
+                if any(tax.price_include for tax in flattened_taxes_before_fp):
+                    taxes_res = flattened_taxes_before_fp.compute_all(
+                        product_price_unit,
+                        quantity=1.0,
+                        currency=currency,
+                        product=product,
+                        partner=partner,
+                        is_refund=is_refund_document,
+                    )
+                    product_price_unit = taxes_res['total_excluded']
+
+                flattened_taxes_after_fp = product_taxes_after_fp._origin.flatten_taxes_hierarchy()
+                if any(tax.price_include for tax in flattened_taxes_after_fp):
+                    taxes_res = flattened_taxes_after_fp.compute_all(
+                        product_price_unit,
+                        quantity=1.0,
+                        currency=currency,
+                        product=product,
+                        partner=partner,
+                        is_refund=is_refund_document,
+                        handle_price_include=False,
+                    )
+                    for tax_res in taxes_res['taxes']:
+                        tax = self.env['account.tax'].browse(tax_res['id'])
+                        if tax.price_include:
+                            product_price_unit += tax_res['amount']
+
+        # Apply currency rate.
+        if currency and currency != company_currency and not self.env.context.get('skip_currency_conversion'):
+            product_price_unit = company_currency._convert(product_price_unit, currency, company, move_date)
```

Comment:
to summarize all my comments above:
```py
@api.model
def _get_default_price_unit(self, company, currency, product, document_date, document_type, is_refund_document=False,
                            product_uom=None, product_currency=None, product_price_unit=None,
                            fiscal_position=None):
    assert document_type

    if product_uom is None:
        product_uom = product.uom_id

    if not product_currency:
        product_currency = company.currency_id

    if product_price_unit is None:
        if document_type == 'sale':
            product_price_unit = product.lst_price
        elif document_type == 'purchase':
            product_price_unit = product.standard_price
        else:
            return 0.0

    if document_type == 'sale':
        product_taxes = product.taxes_id
    else:
        product_taxes = product.supplier_taxes_id
    product_taxes = product_taxes.filtered(lambda x: x.company_id == company)

    # Apply unit of measure.
    ...

    # Apply fiscal position.
    ...

    # Apply currency rate.
    if currency != product_currency:
        product_price_unit = product_currency._convert(product_price_unit, currency, company, document_date)

    return product_price_unit
```

So, in `sale`, you can skip the conversion using the same currency/product_currency